src/storage-backend.js

```diff
@@ -0,0 +1,655 @@
+ *  identify this particular backend.  This should be unique to the
+ *  billing/usage grouping.  For example, we could have two S3Backend instances
+ *  for us-west-2, but because those two backends do work in the same region
+ *  and have the same billing impact they share an id
+ *  - allowedPatterns: each URL passed into the put() method to insert into our
+ *  cache must match at least one of these regular expressions
+ *  - memcachedTTL: the amount of time that a cache entry is put into memcached for
+ *  - memcached: a reference to the memcached object we will use to cache our
+ *  backend addresses
+ *  - sqs: a reference to an sqs object which will be used to listen for
+ *  incoming copy requests as well as to send out completion messages
+ *  - redirectLimit: maximum number of redirects allowed (default 30)
+ *  - ensureSSL: ensure that all redirects are to SSL addresses (default true,
+ *  must be boolean)
+ */
+class StorageBackend {
+  constructor(config) {
+    this.config = config;
+
+    // A nice identifier for error messages
+    assert(this.config.id, 'Missing storage backend ID');
+    this.id = this.config.id;
+    this.copyRequestQueueName = this.id + '_copy_requests';
+
+    // Allowed Patterns is a list of regular expressions which will operate on decoded
+    // rawUrls to determine if they are allowed in this system
+    assert(this.config.allowedPatterns, 'Must specify allowed patterns URLs');
+    this.allowedPatterns = this.config.allowedPatterns;
+
+    // This is the number of seconds that we should keep the url in cache
+    assert(this.config.memcachedTTL, 'Must specify how long to keep urls in cache');
+    this.memcachedTTL = this.config.memcachedTTL;
+    
+    // A configured Memcached object which will be used to store the cache locations
+    assert(this.config.memcached, 'Must provide a memcached object');
+    this.memcached = this.config.memcached;
+
+    // SQS Instance to use for communications
+    assert(this.config.sqs, 'Must provide an SQS object');
+    this.sqs = this.config.sqs;
+
+    // Maximum number of redirects to follow
+    this.redirectLimit = config.redirectLimit || 30;
+
+    // Should we allow non-SSL links in the redirect chain
+    this.ensureSSL = typeof config.ensureSSL === 'boolean' ? config.ensureSSL : true;
+
+    return this;
+  }
+
+  /**
+   * Because constructors are not able to be asyncronous, we have an async init
+   * method which will be used to do all async setup.  This should be called
+   * regardless of whether this instance will listen for incoming file-put
+   * requests.
+   */
+  async init() {
+    let data = await this.sqs.createQueue({
+      QueueName: this.copyRequestQueueName,
+    }).promise();
+
+    // The SQSConsumer rebinds the handleMessage function to it's own scope, so
+    // I want to make sure that the function has access to this object's
+    // properties.
+    let that = this;
```

Comment:
Why do `that` :)
`.bind()` on an arrow function cannot overwrite `this`... Just make sure you're using an arrow function and all is fine.